kyo-stm/shared/src/main/scala/kyo/STM.scala

```diff
@@ -108,34 +111,97 @@ object STM:
-                                val (locked, unlocked) =
-                                    // Sort references by identity to prevent deadlocks
-                                    log.toSeq.sortBy((ref, _) => ref.hashCode)
-                                        .span((ref, entry) => ref.lock(entry))
-
-                                if unlocked.nonEmpty then
-                                    // Failed to acquire some locks - rollback and retry
-                                    locked.foreach((ref, entry) => ref.unlock(entry))
-                                    Abort.fail(FailedTransaction(frame))
-                                else
-                                    // Successfully locked all references - commit changes
-                                    locked.foreach((ref, entry) => ref.commit(tid, entry))
-                                    // Release all locks
-                                    locked.foreach((ref, entry) => ref.unlock(entry))
+                        Var.runWith(TRefLog.empty)(v) { (log, result) =>
+                            val logMap = log.toMap
+                            logMap.size match
+                                case 0 =>
+                                    // Nothing to commit
                                     result
-                                end if
-                            }
+                                case 1 =>
+                                    // Fast-path for a single ref
+                                    IO.Unsafe {
+                                        val (ref, entry) = logMap.head
+                                        // No need to pre-validate since `lock` validates and
+                                        // there's a single ref
+                                        if ref.lock(entry) then
+                                            ref.commit(tid, entry)
+                                            ref.unlock(entry)
+                                            result
+                                        else
+                                            Abort.fail(FailedTransaction(frame))
+                                        end if
+                                    }
+                                case size =>
+                                    // Commit multiple refs
+                                    IO.Unsafe {
+                                        // Flattened representation of the log
+                                        val array = new Array[Any](size * 2)
+
+                                        try
+                                            def fail = throw new FailedTransaction(frame)
+
+                                            var i = 0
+                                            // Pre-validate and dump the log to the flat array
+                                            logMap.foreachEntry { (ref, entry) =>
```

Comment:
I thought Scala's `Map` didn't have a method to traverse the entries without allocating tuples but I've found `foreachEntry` that provides exactly that! Since there's no `forallEntry`, I'm using exception throwing as an efficient way to short circuit